MPP-3284: Log details of region determination

jwhitlock opened the pull request:
This PR addresses MPP-3284. It updates `_get_cc_from_request` in `privaterelay/utils.py` to log the details of the user's country code selection. The details are extra fields on the "eventsinfo" log message "region_details":
    - ~`selected`~ `region_method` - which method was selected, one of "cdn", "accept_lang", "fallback" (US)
    - `region` - the region code returned from the function
    - `cdn_region` - The region code returned from the `X-Client-Region` header, or omitted if none
    - `accept_lang` - The user's `Accept-Language` header
    - `accept_lang_region` - The region code determined from the `Accept-Language` header, or omitted if none

This log message is omitted once per request, by adding a flag on the request.

I originally wanted to do this logging in the `ResponseMetrics` middleware, but that had some problems. See PR #3983. I may pick up that idea again in the future.

## How to test:

Optional test steps to see the log message:

* Run `./manage.py runserver`
* Visit http://127.0.0.1:8000/api/v1/runtime_data/ . Among others, you'll see a log entry like:
   ```json
  {"Timestamp": 1696881837835036928, "Type": "eventsinfo", "Logger": "fx-private-relay", "Hostname": "jwhitlock-MB-M2.local", "EnvVersion": "2.0", "Severity": 6, "Pid": 58553, "Fields": {"accept_lang": "en-US, en", "accept_lang_region": "US", "region_method": "accept_lang", "region": "US", "msg": "region_details"}}
   ```

- [x] I've added a unit test to test for potential regressions of this bug.
- [x] Commits in this PR are minimal and [have descriptive commit messages](https://chris.beams.io/posts/git-commit/).

groovecoder commented on privaterelay/utils.py:
```diff
@@ -55,11 +58,37 @@ def get_countries_info_from_lang_and_mapping(
-        return request.headers["X-Client-Region"].upper()
+        cdn_region = region = request.headers["X-Client-Region"].upper()
+        log_data["cdn_region"] = cdn_region
+        log_data["selected"] = "cdn"
```
suggestion (non-blocking): `selected` is ambiguous. When I first saw it I expected `selected` to mean "Which region was selected" - e.g., `selected: DE`. Can we change it to `region_selection_method` to be more explicit?

jwhitlock replied:
is `region_method` OK? I was mostly avoiding `method`, which is ambiguous with HTTP request method (`POST`, `GET`, etc)

groovecoder commented on privaterelay/utils.py:
```diff
@@ -55,11 +58,37 @@ def get_countries_info_from_lang_and_mapping(
+    log_data["region"] = region
+
+    # MPP-3284: Log details of region selection. Only log once per request, since some
+    # endpoints, like /api/v1/runtime_data, call this multiple times.
```
praise: cool, thanks for catching and preventing the double-log.

groovecoder reviewed (changes requested):
change (blocking): We seem to be missing a test for `test_get_countries_info_cdn_and_accept_language_uses_cdn`? I.e., to test that a request with BOTH `X-Client-Region` and `Accept-Language` header values will use the region from `X-Client-Region`?

jwhitlock commented:
Thanks for the review, @groovecoder! I've changed the name from `selected` to `region_method` (code and tests) and added the test for both headers set (test only).

groovecoder reviewed (approved):
Looks good! Thanks!
